modules/gdscript/gdscript_tokenizer.cpp

```diff
@@ -162,6 +162,24 @@ const char *GDScriptTokenizer::Token::get_name() const {
 	return token_names[type];
 }
 
+bool GDScriptTokenizer::Token::can_precede_bin_op() const {
+	switch (type) {
+		case IDENTIFIER:
+		case LITERAL:
+		case SELF:
+		case BRACKET_CLOSE:
+		case BRACE_CLOSE:
+		case PARENTHESIS_CLOSE:
+		case CONST_PI:
+		case CONST_TAU:
+		case CONST_INF:
+		case CONST_NAN:
+			return true;
+		default:
+			return false;
+	}
+}
+
```

Comment:
I encountered the same problem in the highlighter, and the line of code for it winded up being

```gdscript
if (word == "true" || word == "false" || word == "null" || word == "PI" || word == "TAU" || word == "INF" || word == "NAN" || word == "self" || word == "super" || !reserved_keywords.has(word)
```
not so much in terms of correctness of syntax, but with the rule that if the previous word represents a value, it should be followed by a binary OP. For `super` it was more so because I decided to treat `.` like an operator too as it was helpful to determine when to highlight it like a number.

Maybe there's something to learn from this implementation, and maybe this function in the tokenizer can eventually be used in the highlighter instead of that long condition?

Reply:
I've also noticed that the highlighter doesn't properly handle newlines and line continuations (for example, the `&&` operator doesn't always highlight correctly). Note that after the `NEWLINE` token, `+` must be treated as unary operator or the start of a numeric literal. But if it is a line continuation or multiline mode is enabled, then `+` can be binary operator.

I think for highlighter you can just keep a parentheses stack and flag if the current line is a continuation. Although with lambdas it can be difficult.

Reply:
The highlighter works line by line rn and idk how to fix it. This also messes with the highlighting of much more than just operators that can be unary or binary.

What I'm saying, maybe in the future we can use the tokenizer's info for this too. And also, some of the checks aren't consistent between our implementations, so I wanted to explain my motivation for them.